### PR TITLE
effects/effectchainslot: Add COs to retrieve number of chain presets

### PR DESCRIPTION
**Pull request by Holzhaus**

### CODE REVIEW

File: src/effects/effectchainslot.cpp [hunk 1]
```diff
@@ -40,6 +40,29 @@ EffectChainSlot::EffectChainSlot(const QString& group,
...
+    m_pControlNumQuickEffectChainPresetsAvailable = std::make_unique<ControlObject>(
+            ConfigKey(m_group, "num_quickeffectchainpresetsavailable"));
+    m_pControlNumQuickEffectChainPresetsAvailable->set(
+            m_pChainPresetManager->numQuickEffectPresets());
+    m_pControlNumQuickEffectChainPresetsAvailable->setReadOnly();
+    connect(m_pChainPresetManager.get(),
+            &EffectChainPresetManager::quickEffectChainPresetListUpdated,
+            [this] {
+                m_pControlNumQuickEffectChainPresetsAvailable->forceSet(
+                        m_pChainPresetManager->numQuickEffectPresets());
```
Comment (Be-ing) [inline src/effects/effectchainslot.cpp]: I don't think we should have a separate CO for the QuickEffect presets list. Wrap this line in a protected function and override it in QuickEffectChainSlot.

Comment (Holzhaus) [reply src/effects/effectchainslot.cpp]: Done.

File: src/effects/effectchainslot.cpp [hunk 2]
```diff
@@ -40,6 +40,19 @@ EffectChainSlot::EffectChainSlot(const QString& group,
...
+    m_pControlNumPresetsAvailable->set(m_pChainPresetManager->numPresets());
+    m_pControlNumPresetsAvailable->setReadOnly();
+    connect(m_pChainPresetManager.get(),
+            &EffectChainPresetManager::effectChainPresetListUpdated,
+            this,
+            &EffectChainSlot::slotPresetListUpdated);
+    connect(m_pChainPresetManager.get(),
+            &EffectChainPresetManager::quickEffectChainPresetListUpdated,
+            this,
+            &EffectChainSlot::slotPresetListUpdated);
```
Comment (Be-ing) [inline src/effects/effectchainslot.cpp:54]: EffectChainPresetManager::effectChainPresetListUpdated should not be connected in QuickEffectChainSlot and EffectChainPresetManager::quickEffectChainPresetListUpdated should *only* be connected in QuickEffectChainSlot

File: src/effects/effectchainslot.cpp [hunk 3]
```diff
@@ -40,6 +40,19 @@ EffectChainSlot::EffectChainSlot(const QString& group,
             ConfigKey(m_group, "num_effectslots"));
     m_pControlNumEffectSlots->setReadOnly();
 
+    m_pControlNumPresetsAvailable = std::make_unique<ControlObject>(
+            ConfigKey(m_group, "num_presetsavailable"));
+    m_pControlNumPresetsAvailable->set(m_pChainPresetManager->numPresets());
```
Comment (Be-ing) [inline src/effects/effectchainslot.cpp:45]:
```suggestion
    m_pControlNumPresetsAvailable->set(numPresets());
```

File: src/effects/specialeffectchainslots.cpp [hunk 4]
```diff
@@ -114,6 +114,7 @@ QuickEffectChainSlot::QuickEffectChainSlot(const QString& group,
         addEffectSlot(formatEffectSlotGroup(group, i));
         m_effectSlots.at(i)->setEnabled(true);
     }
+    m_pControlNumPresetsAvailable->set(m_pChainPresetManager->numQuickEffectPresets());
```
Comment (Be-ing) [inline src/effects/specialeffectchainslots.cpp:117]: redundant if the base class uses `numPresets()` for this

File: src/effects/effectchainslot.cpp (hunk 3 above)
Comment (Holzhaus) [reply src/effects/effectchainslot.cpp:45]: We mustn't call virtual methods in constructors.

File: src/effects/specialeffectchainslots.cpp (hunk 4 above)
Comment (Holzhaus) [reply src/effects/specialeffectchainslots.cpp:117]: No. When the base class constructor runs, the subclass doesn't exist yet which means the `numPresets` implementation of the base class is called.

File: src/effects/effectchainslot.cpp (hunk 2 above)
Comment (Holzhaus) [reply src/effects/effectchainslot.cpp:54]: I have no idea how to implement this without calling virtual methods from the constructor.

Comment (Be-ing) [reply src/effects/effectchainslot.cpp:54]: You could delete the connection for the QuickEffects here then in the QuickEffectChainSlot constructor, disconnect from EffectChainPresetManager::effectChainPresetListUpdated and connect to EffectChainPresetManager::quickEffectChainPresetListUpdated

Comment (Holzhaus) [reply src/effects/effectchainslot.cpp:54]: Tried that here: https://github.com/Holzhaus/mixxx/commit/0faf9c22595e0c55a579e6f0f5383a9a9b3354c9

Unfortunately it doesn't work properly (num_presetsavailable is always 0 in quickeffectchainslots). No idea why.

Comment (Be-ing) [reply src/effects/effectchainslot.cpp:54]: The problem was that the quickEffectChainPresetListUpdated signal was not being emitted when effects.xml was empty. I fixed this [here](https://github.com/mixxxdj/mixxx/pull/2618/commits/8503850dfd116f172be8b1e2e9a1cafc2cef18bc#diff-e265e2ac4c841e46240a9480893b1ec3R483) and squashed it together with this PR into one commit.

**Pull request closed without merging**